Aggregation store - Could not instantiate specified check

<!--- Provide a general summary of the issue in the Title above -->

In release 6.0.3, `user check` does not seem to work correctly but in previous releases including 6.0.2 and 6.0.1, the same `user check` works perfectly.


## Context

### I have the following setup
- An independent authentication service using Spring Security that returns a token with a role for a successful authentication
- A business service based on Elide 6.0.3, creating analytic models through hjson.
- `User check` in the business service looks like this:

```
@SecurityCheck(value = "admin")
public class IsAdmin extends UserCheck {

    @Override
    public boolean ok(User user) {
        return user.isInRole("ADMIN");
    }

}
```

### When I start the business service, an exception with below stack trace is thrown

```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.yahoo.elide.Elide]: Factory method 'initializeElide' threw exception; nested exception is java.lang.IllegalArgumentException: Could not instantiate specified check 'admin'.
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)
	... 33 common frames omitted
Caused by: java.lang.IllegalArgumentException: Could not instantiate specified check 'admin'.
	at com.yahoo.elide.core.dictionary.EntityDictionary.lambda$getCheck$5(EntityDictionary.java:381)
	at java.base/java.util.Map.computeIfAbsent(Map.java:1002)
	at com.google.common.collect.Synchronized$SynchronizedMap.computeIfAbsent(Synchronized.java:1185)
	at com.yahoo.elide.core.dictionary.EntityDictionary.getCheck(EntityDictionary.java:377)
	at com.yahoo.elide.datastores.aggregation.AggregationDataStore.lambda$validateExpression$11(AggregationDataStore.java:184)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1694)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at com.yahoo.elide.datastores.aggregation.AggregationDataStore.validateExpression(AggregationDataStore.java:183)
	at com.yahoo.elide.datastores.aggregation.AggregationDataStore.validateModelExpressionChecks(AggregationDataStore.java:151)
	at com.yahoo.elide.datastores.aggregation.AggregationDataStore.lambda$populateEntityDictionary$3(AggregationDataStore.java:93)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1694)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at com.yahoo.elide.datastores.aggregation.AggregationDataStore.populateEntityDictionary(AggregationDataStore.java:91)
	at com.yahoo.elide.datastores.multiplex.MultiplexManager.populateEntityDictionary(MultiplexManager.java:68)
	at com.yahoo.elide.core.datastore.inmemory.InMemoryDataStore.populateEntityDictionary(InMemoryDataStore.java:27)
	at com.yahoo.elide.Elide.doScans(Elide.java:121)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration.initializeElide(ElideAutoConfiguration.java:191)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$e63f001a.CGLIB$initializeElide$6(<generated>)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$e63f001a$$FastClassBySpringCGLIB$$5e6fdcf8.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$e63f001a.initializeElide(<generated>)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 34 common frames omitted
Caused by: java.lang.ClassNotFoundException: admin
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:606)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:168)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:377)
	at com.yahoo.elide.core.dictionary.EntityDictionary.lambda$getCheck$5(EntityDictionary.java:379)
	... 74 common frames omitted

```


## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Elide version used: 6.0.3
* Spring boot version used: 2.6.1
* Environment name and version (Java 1.8.0_152): 15
* Operating System and version: MacOS
* Link to your project: N/A

Added the check manually to dictionary, then face another exception with below stack trace:

```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.yahoo.elide.modelconfig.DynamicConfiguration]: Factory method 'buildDynamicConfiguration' threw exception; nested exception is java.lang.IllegalStateException: Found undefined security checks: [admin]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)
	... 33 common frames omitted
Caused by: java.lang.IllegalStateException: Found undefined security checks: [admin]
	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateChecks(DynamicConfigValidator.java:645)
	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateTableConfig(DynamicConfigValidator.java:592)
	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateConfigs(DynamicConfigValidator.java:216)
	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validate(DynamicConfigValidator.java:177)
	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.readAndValidateConfigs(DynamicConfigValidator.java:190)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration.buildDynamicConfiguration(ElideAutoConfiguration.java:111)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d.CGLIB$buildDynamicConfiguration$2(<generated>)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d$$FastClassBySpringCGLIB$$85eaaed0.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d.buildDynamicConfiguration(<generated>)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 34 common frames omitted
```

I looked into the `DynamicValidator` and this exception is quite strange to me as the check is not a generated one and the check is not extend [RoleMemberCheck.](https://github.com/yahoo/elide/blob/master/elide-core/src/main/java/com/yahoo/elide/core/security/checks/prefab/Role.java#L38-L48) either.

Maybe the check doesn’t have equals and hash code defined and it is getting
lost in the map?

Aaron

On Wed, Dec 8, 2021 at 5:23 AM Thai Ngo ***@***.***> wrote:

> Added the check manually to dictionary, then face another exception with
> below stack trace:
>
> Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.yahoo.elide.modelconfig.DynamicConfiguration]: Factory method 'buildDynamicConfiguration' threw exception; nested exception is java.lang.IllegalStateException: Found undefined security checks: [admin]
> 	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
> 	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)
> 	... 33 common frames omitted
> Caused by: java.lang.IllegalStateException: Found undefined security checks: [admin]
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateChecks(DynamicConfigValidator.java:645)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateTableConfig(DynamicConfigValidator.java:592)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateConfigs(DynamicConfigValidator.java:216)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validate(DynamicConfigValidator.java:177)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.readAndValidateConfigs(DynamicConfigValidator.java:190)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration.buildDynamicConfiguration(ElideAutoConfiguration.java:111)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d.CGLIB$buildDynamicConfiguration$2(<generated>)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d$$FastClassBySpringCGLIB$$85eaaed0.invoke(<generated>)
> 	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
> 	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d.buildDynamicConfiguration(<generated>)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
> 	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
> 	... 34 common frames omitted
>
> I looked into the DynamicValidator and this exception is quite strange to
> me as the check is not a generated one and the check is not extend
> RoleMemberCheck.
> <https://github.com/yahoo/elide/blob/master/elide-core/src/main/java/com/yahoo/elide/core/security/checks/prefab/Role.java#L38-L48>
> either.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/yahoo/elide/issues/2438#issuecomment-988727554>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVH6QJX5OQ3PSYDRYSQMNTUP4533ANCNFSM5JRXCB3Q>
> .
>

> Added the check manually to dictionary, then face another exception with below stack trace:
> 
> ```
> Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.yahoo.elide.modelconfig.DynamicConfiguration]: Factory method 'buildDynamicConfiguration' threw exception; nested exception is java.lang.IllegalStateException: Found undefined security checks: [admin]
> 	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
> 	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)
> 	... 33 common frames omitted
> Caused by: java.lang.IllegalStateException: Found undefined security checks: [admin]
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateChecks(DynamicConfigValidator.java:645)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateTableConfig(DynamicConfigValidator.java:592)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validateConfigs(DynamicConfigValidator.java:216)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.validate(DynamicConfigValidator.java:177)
> 	at com.yahoo.elide.modelconfig.validator.DynamicConfigValidator.readAndValidateConfigs(DynamicConfigValidator.java:190)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration.buildDynamicConfiguration(ElideAutoConfiguration.java:111)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d.CGLIB$buildDynamicConfiguration$2(<generated>)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d$$FastClassBySpringCGLIB$$85eaaed0.invoke(<generated>)
> 	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
> 	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
> 	at com.yahoo.elide.spring.config.ElideAutoConfiguration$$EnhancerBySpringCGLIB$$c13a642d.buildDynamicConfiguration(<generated>)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
> 	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
> 	... 34 common frames omitted
> ```
> 
> I looked into the `DynamicValidator` and this exception is quite strange to me as the check is not a generated one (and does not extend [RoleMemberCheck.](https://github.com/yahoo/elide/blob/master/elide-core/src/main/java/com/yahoo/elide/core/security/checks/prefab/Role.java#L38-L48) either)

You can reference both your dynamic role based checks and static class checks in HJSON configuration.  The validator checks both when ensuring that all security checks are present.

--- a/elide-core/src/main/java/com/yahoo/elide/Elide.java
+++ b/elide-core/src/main/java/com/yahoo/elide/Elide.java
@@ -119,6 +119,7 @@ public class Elide {
             elideSettings.getSerdes().forEach((type, serde) -> registerCustomSerde(type, serde, type.getSimpleName()));
             registerCustomSerde();
 
+            //Scan for security checks prior to populating data stores in case they need them.
             elideSettings.getDictionary().scanForSecurityChecks();
 
             this.dataStore.populateEntityDictionary(elideSettings.getDictionary());

--- a/elide-core/src/main/java/com/yahoo/elide/Elide.java
+++ b/elide-core/src/main/java/com/yahoo/elide/Elide.java
@@ -119,9 +119,9 @@ public class Elide {
             elideSettings.getSerdes().forEach((type, serde) -> registerCustomSerde(type, serde, type.getSimpleName()));
             registerCustomSerde();
 
-            this.dataStore.populateEntityDictionary(elideSettings.getDictionary());
-
             elideSettings.getDictionary().scanForSecurityChecks();
+
+            this.dataStore.populateEntityDictionary(elideSettings.getDictionary());
             initialized = true;
         }
     }

--- a/elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/integration/AggregationDataStoreIntegrationTest.java
+++ b/elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/integration/AggregationDataStoreIntegrationTest.java
@@ -37,8 +37,6 @@ import com.yahoo.elide.core.security.checks.Check;
 import com.yahoo.elide.core.security.checks.prefab.Role;
 import com.yahoo.elide.core.utils.DefaultClassScanner;
 import com.yahoo.elide.datastores.aggregation.AggregationDataStore;
-import com.yahoo.elide.datastores.aggregation.checks.OperatorCheck;
-import com.yahoo.elide.datastores.aggregation.checks.VideoGameFilterCheck;
 import com.yahoo.elide.datastores.aggregation.framework.AggregationDataStoreTestHarness;
 import com.yahoo.elide.datastores.aggregation.framework.SQLUnitTest;
 import com.yahoo.elide.datastores.aggregation.metadata.enums.TimeGrain;

--- a/elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/integration/AggregationDataStoreIntegrationTest.java
+++ b/elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/integration/AggregationDataStoreIntegrationTest.java
@@ -114,8 +114,6 @@ public class AggregationDataStoreIntegrationTest extends GraphQLIntegrationTest 
                 @Override
                 protected void configure() {
                     Map<String, Class<? extends Check>> map = new HashMap<>(TestCheckMappings.MAPPINGS);
-                    map.put(OperatorCheck.OPERTOR_CHECK, OperatorCheck.class);
-                    map.put(VideoGameFilterCheck.NAME_FILTER, VideoGameFilterCheck.class);
                     EntityDictionary dictionary = EntityDictionary.builder().checks(map).build();
 
                     VALIDATOR.getElideSecurityConfig().getRoles().forEach(role ->

